File: src/main/java/AoC2019_17.java
import static com.github.pareronia.aoc.AssertUtils.assertTrue;
import static com.github.pareronia.aoc.IntegerSequence.Range.range;
import static com.github.pareronia.aoc.StringOps.splitLines;
import static com.github.pareronia.aoc.Utils.concatAll;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toSet;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.function.Predicate;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.ListUtils;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Turn;
import com.github.pareronia.aoc.intcode.IntCode;
import com.github.pareronia.aoc.solution.Logger;
import com.github.pareronia.aoc.solution.LoggerEnabled;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2019_17 extends SolutionBase<List<Long>, Integer, Integer> {
    
    private static final char SCAFFOLD = '#';
    private static final char NEWLINE = '\n';
    
    private AoC2019_17(final boolean debug) {
        super(debug);
    }

    public static AoC2019_17 create() {
        return new AoC2019_17(false);
    }

    public static AoC2019_17 createDebug() {
        return new AoC2019_17(true);
    }

    @Override
    protected List<Long> parseInput(final List<String> inputs) {
        return IntCode.parse(inputs.get(0));
    }
    
    @Override
    public Integer solvePart1(final List<Long> program) {
        final CharGrid grid = new GridBuilder().build(
            new IntCodeComputer(program, this.debug).runCamera());
        log(grid);
        return grid.getAllEqualTo(SCAFFOLD)
            .filter(cell -> grid.getCapitalNeighbours(cell)
                        .allMatch(n -> grid.getValue(n) == SCAFFOLD))
            .mapToInt(cell -> cell.getRow() * cell.getCol())
            .sum();
    }

    @Override
    public Integer solvePart2(final List<Long> program) {
        final IntCodeComputer computer = new IntCodeComputer(program, this.debug);
        final CharGrid grid = new GridBuilder().build(computer.runCamera());
        final List<String> input = new PathFinder(grid, this.logger).createAsciiInput(5, 3);
        return computer.runRobot(input).getLast().intValue();
    }

    @Override
    protected void samples() {
        assert new PathFinder(CharGrid.from(splitLines(TEST)), new Logger(true))
            .createAsciiInput(3, 2)
            .equals(List.of("A,B,C,B,A,C", "R,8,R,8", "R,4,R,4,R,8", "L,6,L,2", "n"));
    }

    public static void main(final String[] args) throws Exception {
        AoC2019_17.create().run();
    }
    
    private static final String TEST = """
            #######...#####
            #.....#...#...#
            #.....#...#...#
            ......#...#...#
            ......#...###.#
            ......#.....#.#
            ^########...#.#
            ......#.#...#.#
            ......#########
            ........#...#..
            ....#########..
            ....#...#......
            ....#...#......
            ....#...#......
            ....#####......
            """;
    
    private static final class GridBuilder {

        public CharGrid build(final Deque<Long> output) {
            return CharGrid.from(asStrings(output));
        }
        
        private List<String> asStrings(final Deque<Long> output) {
            final List<String> strings = new ArrayList<>();
            final StringBuilder sb = new StringBuilder();
            while (!output.isEmpty()) {
                final char out = (char) (long) output.pop();
                if (out == NEWLINE && sb.length() > 0) {
                    strings.add(sb.toString());
                    sb.setLength(0);
                } else {
                    sb.append(out);
                }
            }
            return strings;
        }
    }
    
    record Command(char letter, int count) {
        
        @Override
        public String toString() {
            if (this.count == 1) {
                return String.valueOf(this.letter);
            } else {
                return String.format("%s(%d)", this.letter, this.count);
            }
        }
    }
    
    private static final class PathFinder implements LoggerEnabled {
        private final CharGrid grid;
        private final Logger logger;
        
        public PathFinder(final CharGrid grid, final Logger logger) {
            this.grid = grid;
            this.logger = logger;
        }

        @Override
        public Logger getLogger() {
            return this.logger;
        }

        public List<Cell> findPath() {
            final Cell robot = grid.findAllMatching(Direction.CAPITAL_ARROWS::contains)
                .findFirst().orElseThrow();
            final Cell start = robot;
            final Cell end = grid.getAllEqualTo(SCAFFOLD)
                .filter(cell -> grid.getCapitalNeighbours(cell)
                    .filter(n -> grid.getValue(n) == SCAFFOLD || n.equals(robot))
                    .count() == 1)
                .findFirst().orElseThrow();
            final Deque<Cell> path = new ArrayDeque<>();
            final DFS dfs = new DFS(grid, path, start, end);
            dfs.dfs();
            return path.stream().collect(toList());
        }
        
        public List<Command> toCommands(final List<Direction> moves) {
            final List<List<Command>> commands = new ArrayList<>();
            final Deque<Command> curr = new ArrayDeque<>();
            range(moves.size() - 1).forEach(i -> {
                final Command last = curr.peekLast();
                if (moves.get(i) == moves.get(i + 1)){
                    curr.addLast(new Command(last.letter, 1));
                } else {
                    final Turn turn = Turn.fromDirections(
                            moves.get(i), moves.get(i + 1));
                    if (last != null) {
                        commands.add(curr.stream().collect(toList()));
                        curr.clear();
                    }
                    curr.addLast(new Command(turn.getLetter().get(), 1));
                }
            });
            commands.add(curr.stream().collect(toList()));
            return commands.stream()
                    .map(lst -> new Command(lst.get(0).letter, lst.size()))
                    .toList();
        }
        
        public List<String> createAsciiInput(
                final int maxSize, final int minRepeats
        ) {
            final List<Cell> path = this.findPath();
            log("Path: " + path);
            final List<Direction> moves = concatAll(
                List.of(Direction.fromChar(grid.getValue(path.get(0)))),
                range(path.size() - 1).intStream()
                    .mapToObj(i1 -> path.get(i1).to(path.get(i1 + 1)))
                    .toList());
            log("Moves: " + moves);
            final List<Command> commands = this.toCommands(moves);
            log("Commands: " + commands);
            List<Command> lst = new ArrayList<>(commands);
            final Map<String, List<Command>> map = new HashMap<>();
            for (final String x : List.of("A", "B", "C")) {
                for (int i = maxSize; i >= 2; i--) {
                    if (i > lst.size()) {
                        continue;
                    }
                    final List<Integer> idxs = ListUtils.indexesOfSubList(
                                                    lst, lst.subList(0, i));
                    if (idxs.size() < minRepeats) {
                        continue;
                    } else {
                        map.put(x, new ArrayList<>(lst.subList(0, i)));
                        lst = ListUtils.subtractAll(lst, lst.subList(0, i));
                        break;
                    }
                }
            }
            log(map);
            assertTrue(map.size() == 3, () -> "could not find functions");
            final List<String> main = new ArrayList<>();
            lst = new ArrayList<>(commands);
            int cnt = 0;
            while (!lst.isEmpty()) {
                assertTrue(cnt < 100, () -> "infinite loop");
                for (final Entry<String, List<Command>> func : map.entrySet()) {
                    if (Collections.indexOfSubList(lst, func.getValue()) == 0) {
                        main.add(func.getKey());
                        lst = lst.subList(func.getValue().size(), lst.size());
                        log(lst);
                        break;
                    }
                }
                cnt++;
            }
            log(main);
            final List<String> input = new ArrayList<>();
            input.add(main.stream().collect(joining(",")));
            List.of("A", "B", "C").stream()
                .map(x -> map.get(x).stream()
                        .map(c -> "%s,%d".formatted(c.letter, c.count))
                        .collect(joining(",")))
                .forEach(input::add);
            input.add("n");
            log(input);
            return input;
        }
        
        private static final class DFS {
            private final CharGrid grid;
            private final Deque<Cell> path;
            private final Set<State> seen;
            private final Predicate<State> isEnd;
            private State state;
            
            public DFS(
                final CharGrid grid,
                final Deque<Cell> path,
                final Cell start,
                final Cell end
            ) {
                this.grid = grid;
                this.path = path;
                this.seen = new HashSet<>();
                this.isEnd = state -> state.curr.equals(end);
                this.state = new State(start, start);
                this.path.addLast(start);
                this.seen.add(this.state);
            }
            
            public boolean dfs() {
                if (isEnd.test(this.state)) {
                    return true;
                }
                final Set<Cell> scaffolds = grid.getCapitalNeighbours(this.state.curr)
                    .filter(n -> grid.getValue(n) == SCAFFOLD)
                    .collect(toSet());
                Set<Cell> adjacent;
                if (scaffolds.size() == 4) {
                    adjacent = scaffolds.stream()
                            .filter(s -> s.getRow() == this.state.prev.getRow()
                            || s.getCol() == this.state.prev.getCol()).collect(toSet());
                } else {
                    adjacent = scaffolds;
                }
                for (final Cell cell : adjacent) {
                    final State newState = new State(this.state.curr, cell);
                    if (seen.contains(newState)) {
                        continue;
                    }
                    final State oldState = this.state;
                    this.state = newState;
                    path.addLast(cell);
                    seen.add(this.state);
                    if (dfs()) {
                        return true;
                    } else {
                        path.removeLast();
                        seen.remove(this.state);
                        this.state = oldState;
                    }
                }
                return false;
            }
            
            record State(Cell prev, Cell curr) {}
        }
    }
    
    record IntCodeComputer(List<Long> program, boolean debug) {
        
        public Deque<Long> runCamera() {
            final IntCode intCode = new IntCode(this.program, this.debug);
            final Deque<Long> input = new ArrayDeque<>();
            final Deque<Long> output = new ArrayDeque<>();
            intCode.run(input, output);
            return output;
        }
        
        public Deque<Long> runRobot(final List<String> commands) {
            final List<Long> newProgram = concatAll(
                    List.of(2L),
                    this.program.subList(1, this.program.size()));
            final IntCode intCode = new IntCode(newProgram, this.debug);
            final Deque<Long> input = new ArrayDeque<>();
            final Deque<Long> output = new ArrayDeque<>();
            commands.forEach(s -> (s + NEWLINE).chars()
                    .mapToLong(Long::valueOf)
                    .forEach(input::addLast));
            intCode.run(input, output);
            return output;
        }
    }
}

File: src/main/python/AoC2019_17.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 17
#

from __future__ import annotations

import sys
from typing import Callable
from typing import NamedTuple

from aoc.collections import index_of_sublist
from aoc.collections import indexes_of_sublist
from aoc.collections import subtract_all
from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import log
from aoc.geometry import Direction
from aoc.geometry import Turn
from aoc.grid import Cell
from aoc.intcode import IntCode

NEWLINE = "\n"
SCAFFOLD = "#"
FUNCTIONS = ["A", "B", "C"]
Input = list[int]
Output1 = int
Output2 = int

TEST = """\
#######...#####
#.....#...#...#
#.....#...#...#
......#...#...#
......#...###.#
......#.....#.#
^########...#.#
......#.#...#.#
......#########
........#...#..
....#########..
....#...#......
....#...#......
....#...#......
....#####......
"""


class IntCodeComputer(NamedTuple):
    program: list[int]

    def run_camera(self) -> list[int]:
        intcode = IntCode(self.program)
        out = list[int]()
        intcode.run([], out)
        return out

    def run_robot(self, commands: list[str]) -> list[int]:
        intcode = IntCode([2] + self.program[1:])
        input = [ord(ch) for command in commands for ch in command + NEWLINE]
        output = list[int]()
        intcode.run(input, output)
        return output


class CameraFeed(NamedTuple):
    scaffolds: set[Cell]
    robot_cell: Cell
    robot_dir: Direction

    @classmethod
    def ascii_to_strings(cls, ascii: list[int]) -> list[str]:
        strings: list[str] = []
        sb: list[str] = []
        while not len(ascii) == 0:
            o = chr(ascii.pop(0))
            if o == NEWLINE and len(sb) > 0:
                strings.append("".join(sb))
                sb = []
            else:
                sb.append(o)
        return strings

    @classmethod
    def from_ascii(cls, ascii: list[int]) -> CameraFeed:
        strings = CameraFeed.ascii_to_strings(ascii)
        return CameraFeed.from_strings(strings)

    @classmethod
    def from_strings(cls, strings: list[str]) -> CameraFeed:
        arrows = Direction.capital_arrows()
        scaffolds = set[Cell]()
        log((len(strings), len(strings[0])))
        for r in range(len(strings)):
            for c in range(len(strings[0])):
                val = strings[r][c]
                if val in arrows:
                    robot_cell = Cell(r, c)
                    robot_dir = Direction.from_str(val)
                elif val == SCAFFOLD:
                    scaffolds.add(Cell(r, c))
        return CameraFeed(scaffolds, robot_cell, robot_dir)


class Command(NamedTuple):
    letter: str
    cnt: int

    def __str__(self) -> str:
        return self.letter if self.cnt == 1 else f"{self.letter}({self.cnt})"


class PathFinder:
    def __init__(self, feed: CameraFeed) -> None:
        self.feed = feed

    class State(NamedTuple):
        prev: Cell
        curr: Cell

    class DFS:
        def __init__(
            self,
            scaffolds: set[Cell],
            path: list[Cell],
            start: Cell,
            end: Cell,
        ) -> None:
            self.scaffolds = scaffolds
            self.path = path
            self.seen = set[PathFinder.State]()
            self.is_end: Callable[[PathFinder.State], bool] = (
                lambda state: state.curr == end
            )
            self.state = PathFinder.State(start, start)
            self.path.append(start)
            self.seen.add(self.state)

        def dfs(self) -> bool:
            if self.is_end(self.state):
                return True
            sc = {
                n
                for n in self.state.curr.get_capital_neighbours()
                if n in self.scaffolds
            }
            if len(sc) == 4:
                adjacent = {
                    s
                    for s in sc
                    if s.row == self.state.prev.row
                    or s.col == self.state.prev.col
                }
            else:
                adjacent = {_ for _ in sc}
            for cell in adjacent:
                new_state = PathFinder.State(self.state.curr, cell)
                if new_state in self.seen:
                    continue
                old_state = self.state
                self.state = new_state
                self.path.append(cell)
                self.seen.add(self.state)
                if self.dfs():
                    return True
                else:
                    self.path.pop()
                    self.seen.remove(self.state)
                    self.state = old_state
            return False

    def find_path(self) -> list[Cell]:
        start = self.feed.robot_cell
        end = next(
            cell
            for cell in self.feed.scaffolds
            if sum(
                n in self.feed.scaffolds or n == self.feed.robot_cell
                for n in cell.get_capital_neighbours()
            )
            == 1
        )
        path = list[Cell]()
        dfs = PathFinder.DFS(self.feed.scaffolds, path, start, end)
        dfs.dfs()
        return path

    def to_commands(self, moves: list[Direction]) -> list[Command]:
        commands = list[list[Command]]()
        curr = list[Command]()
        for i in range(len(moves) - 1):
            if moves[i] == moves[i + 1]:
                curr.append(Command(curr[-1].letter, 1))
            else:
                turn = Turn.from_directions(moves[i], moves[i + 1])
                if len(curr) > 0:
                    commands.append(curr[:])
                    curr = []
                curr.append(Command(turn.letter, 1))
        commands.append(curr[:])
        return [Command(lst[0].letter, len(lst)) for lst in commands]

    def create_ascii_input(self, max_size: int, min_repeats: int) -> list[str]:
        path = self.find_path()
        # log(path)
        moves = [self.feed.robot_dir] + [
            path[i].to(path[i + 1]) for i in range(len(path) - 1)
        ]
        # log(moves)
        commands = self.to_commands(moves)
        log(", ".join(str(c) for c in commands))
        lst = commands[:]
        d = dict[str, list[Command]]()
        for x in FUNCTIONS:
            for i in range(max_size, 1, -1):
                if i > len(lst):
                    continue
                idxs = indexes_of_sublist(lst, lst[:i])
                if len(idxs) < min_repeats:
                    continue
                else:
                    d[x] = lst[0:i][:]
                    lst = subtract_all(lst, lst[:i])
                    break
        # log(d)
        main = list[str]()
        lst = commands[:]
        while len(lst) > 0:
            for k, v in d.items():
                if index_of_sublist(lst, v) == 0:
                    main.append(k)
                    lst = lst[len(v) :]  # noqa E203
                    break
        log(main)
        ascii_input = (
            [",".join(main)]
            + [
                ",".join(f"{c.letter},{c.cnt}" for c in d[x])
                for x in FUNCTIONS
            ]
            + ["n"]
        )
        log(ascii_input)
        return ascii_input


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [int(_) for _ in list(input_data)[0].split(",")]

    def part_1(self, program: Input) -> Output1:
        feed = CameraFeed.from_ascii(IntCodeComputer(program).run_camera())
        return sum(
            cell.row * cell.col
            for cell in feed.scaffolds
            if all(n in feed.scaffolds for n in cell.get_capital_neighbours())
        )

    def part_2(self, program: Input) -> Output2:
        computer = IntCodeComputer(program)
        path_finder = PathFinder(CameraFeed.from_ascii(computer.run_camera()))
        ascii_input = path_finder.create_ascii_input(max_size=5, min_repeats=3)
        return computer.run_robot(ascii_input)[-1]

    def samples(self) -> None:
        path_finder = PathFinder(CameraFeed.from_strings(TEST.splitlines()))
        assert path_finder.create_ascii_input(max_size=3, min_repeats=2,) == [
            "A,B,C,B,A,C",
            "R,8,R,8",
            "R,4,R,4,R,8",
            "L,6,L,2",
            "n",
        ]


solution = Solution(2019, 17)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
